crates/light-client-prover/src/circuit.rs

```diff
@@ -61,23 +67,69 @@ pub fn run_circuit<DaV: DaVerifier, G: ZkvmGuest>(
         }
     }
 
-    let batch_proof_journals = input.batch_proof_journals;
-    let batch_proof_method_id = input.batch_proof_method_id;
-    // TODO: Test for multiple assumptions to see if the env::verify function does automatic matching between the journal and the assumption or do we need to verify them in order?
-    // https://github.com/chainwayxyz/citrea/issues/1401
-    for journal in batch_proof_journals {
-        G::verify(&journal, &batch_proof_method_id.into()).unwrap();
+    // Deserialize batch proof journals
+    let mut deserialized_outputs: Vec<_> = input
+        .batch_proof_journals
+        .iter()
+        .map(|journal| {
+            G::verify_and_extract_output::<StateTransition<DaV::Spec, [u8; 32]>>(
+                journal,
+                &input.batch_proof_method_id.into(),
+            )
+            .expect("Should have verified and extracted the batch proof")
+        })
+        .collect();
+
+    // If we have a previous light client proof, check they can be chained
+    // If not, skip for now
+    // TODO: Once we have a manually planted light client proof use that and assume prev light client proof always exists
+    // So there will be no need for all these checks
+    if let Some(previous_output) = &deserialized_previous_light_client_proof_journal {
+        // Start with the previous light client proof's final state root
+        let mut current_final_state_root = previous_output.state_root;
+
+        let mut journal_map: HashMap<[u8; 32], usize> = deserialized_outputs
+            .iter()
+            .enumerate()
+            .map(|(index, journal)| (journal.initial_state_root, index))
+            .collect();
+
+        let mut i = 0;
+
+        while i < deserialized_outputs.len() {
+            // Check if the current journal's initial state matches the expected current final state
+            if deserialized_outputs[i].initial_state_root == current_final_state_root {
+                // Update the current final state to this journal's final state
+                current_final_state_root = deserialized_outputs[i].final_state_root;
+                // Move to the next journal
+                i += 1;
+            } else {
+                // Use the map to find the correct journal quickly
+                if let Some(&pos) = journal_map.get(&current_final_state_root) {
+                    if pos > i {
+                        // Swap the found matching journal to the current position
+                        deserialized_outputs.swap(i, pos);
+                        // Update the map to reflect the swap
+                        journal_map.insert(deserialized_outputs[pos].initial_state_root, pos);
+                        journal_map.insert(deserialized_outputs[i].initial_state_root, i);
```

Comment:
I think this is completely unnecessary. We will have at most like 10, maybe 15 batch proofs in the array, and I am pretty sure that iteration over list is going to be much faster for such a small list.

Reply:
I agree as well, I have both versions check https://github.com/chainwayxyz/citrea/pull/1428/commits/791f44138235dc6bd4cb0a15879c47eb7501c3cb 
We will benchmark and see
But personally I would also prefer the the one in https://github.com/chainwayxyz/citrea/pull/1428/commits/791f44138235dc6bd4cb0a15879c47eb7501c3cb because 99% it will be ordered